apps/aecore/lib/aecore/structures/oracle_extend_tx_data.ex

```diff
@@ -0,0 +1,123 @@
+defmodule Aecore.Structures.OracleExtendTxData do
+  @behaviour Aecore.Structures.Transaction
+
+  alias __MODULE__
+  alias Aecore.Oracle.Oracle
+  alias Aecore.Structures.Account
+  alias Aecore.Wallet.Worker, as: Wallet
+
+  require Logger
+
+  @type payload :: %{
+          ttl: non_neg_integer()
+        }
+
+  @type t :: %OracleExtendTxData{
+          ttl: non_neg_integer()
+        }
+
+  defstruct [:ttl]
+  use ExConstructor
+
+  @spec get_chain_state_name() :: :oracles
+  def get_chain_state_name(), do: :oracles
+
+  @spec init(payload()) :: OracleExtendTxData.t()
+  def init(%{ttl: ttl}) do
+    %OracleExtendTxData{ttl: ttl}
+  end
+
+  @spec is_valid?(OracleExtendTxData.t()) :: boolean()
+  def is_valid?(%OracleExtendTxData{ttl: ttl}) do
+    ttl > 0
+  end
+
+  @spec process_chainstate!(
+          OracleExtendTxData.t(),
+          binary(),
+          non_neg_integer(),
+          non_neg_integer(),
+          non_neg_integer(),
+          ChainState.account(),
+          Oracle.registered_oracles()
+        ) :: {ChainState.accounts(), Oracle.registered_oracles()}
+  def process_chainstate!(
+        %OracleExtendTxData{} = tx,
+        sender,
+        fee,
+        nonce,
+        block_height,
+        accounts,
+        %{registered_oracles: registered_oracles} = oracle_state
+      ) do
+    case preprocess_check(
+           tx,
+           sender,
+           Map.get(accounts, sender, Account.empty()),
+           fee,
+           nonce,
+           block_height,
+           registered_oracles
+         ) do
+      :ok ->
+        new_senderount_state =
+          Map.get(accounts, sender, Account.empty())
+          |> deduct_fee(fee)
+
+        updated_accounts_chainstate = Map.put(accounts, sender, new_senderount_state)
+
+        updated_oracle_state =
+          update_in(
+            oracle_state,
+            [:registered_oracles, sender, :tx, Access.key(:ttl), :ttl],
+            &(&1 + tx.ttl)
+          )
+
+        {updated_accounts_chainstate, updated_oracle_state}
+
+      {:error, _reason} = err ->
+        throw(err)
+    end
+  end
+
+  @spec preprocess_check(
+          OracleExtendTxData.t(),
+          Wallet.pubkey(),
+          ChainState.account(),
+          non_neg_integer(),
+          non_neg_integer(),
+          non_neg_integer(),
+          Oracle.registered_oracles()
+        ) :: :ok | {:error, String.t()}
+  def preprocess_check(tx, sender, account_state, fee, nonce, _block_height, registered_oracles) do
+    cond do
```

Comment:
shouldn't this fail if an absolute ttl is below height?

Reply:
The TTL in this transaction is just an integer which is added to the oracle %{ttl: value}

Reply:
okay